src/runtime/ndarray.cc

```diff
@@ -214,18 +220,23 @@ NDArray NDArray::EmptyShared(const std::string &name,
 NDArray NDArray::Empty(std::vector<int64_t> shape,
                        DLDataType dtype,
                        DLContext ctx) {
-  TensorDispatcher* td = TensorDispatcher::Global();
-  if (td->IsAvailable())
-    return td->Empty(shape, dtype, ctx);
-
   NDArray ret = Internal::Create(shape, dtype, ctx);
   // setup memory content
   size_t size = GetDataSize(ret.data_->dl_tensor);
-  size_t alignment = GetDataAlignment(ret.data_->dl_tensor);
-  if (size > 0)
-    ret.data_->dl_tensor.data =
-        DeviceAPI::Get(ret->ctx)->AllocDataSpace(
-            ret->ctx, size, alignment, ret->dtype);
+  TensorDispatcher* td = TensorDispatcher::Global();
+  if (size > 0) {
+    if (td->IsAvailable()) {
+      // when td is avaible, use AllocWorkspace to use PyTorch's allocator
+      ret.data_->dl_tensor.data =
+          DeviceAPI::Get(ret->ctx)->AllocWorkspace(
+              ret->ctx, size, ret->dtype);
+    } else {
+      size_t alignment = GetDataAlignment(ret.data_->dl_tensor);
+      ret.data_->dl_tensor.data =
+          DeviceAPI::Get(ret->ctx)->AllocDataSpace(
+              ret->ctx, size, alignment, ret->dtype);
+    }
+  }
```

Comment:
Same question.

Reply:
In the original DGL without a TensorAdaptor, `AllocDataSpace` is used for allocating persistent memories, e.g., for NDArrays, while `AllocWorkspace` is used for allocating temporary memories, e.g., computing buffers, and it uses a thread-local `WorkspacePool`. If we simply replace `AllocDataSpace` with `AllocWorkspace` here, it may raise an error when freeing, saying that `trying to free things that has not been allocated`. I think this is due to the temporary memory being freed in another thread instead of the one it was allocated in. (Please correct me if I'm wrong.)

In this PR, I redirect `AllocWorkspace` to use PyTorch's allocators for both CPU and GPU via the TensorAdaptor. That's why we use `AllocWorkspace` here when TensorAdaptor is available.

An alternative is to redirect `AllocDataSpace` to PyTorch's allocators too when the TensorAdaptor is available. But I think `AllocDataSpace` is more like a “primitive” and wonder if it's reasonable to do so.

Feel free to tell me if there're better ways to unify the usage.

Reply:
Reverted this change and redirect `AllocDataSpace` to PyTorch's allocators.